### PR TITLE
Adding all init.d scripts generated by chef solo. These are static

### PR DESCRIPTION
**Pull request by zoran-altiscale**
files, so if there is any changes we do in chef solo, it will not work
anymore. The proper way to maintain init.d scripts is to modify
sysconfig files and keep init.d scripts static.

### CODE REVIEW

**Comment by raviprak-altiscale** [comment]
Are there any changes from chef-solo in here? (I do see some references to default['hadoop'] . If that's fine..... +1.

**Comment by zoran-altiscale** [comment]
that's a bug then... let me see again... this was copy from actual machine,
so if anything is not resolved, we already have a bug...

**Comment by zoran-altiscale** [comment]
I see - the reference is in the comment. My suggestion is to move these
options to sysconfig as config can and should be managed by chef.

File: etc/init.d/hadoop_datanode [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+
+if [ "x" == "${JAVA_HOME}x" ]
+then
+    echo "JAVA_HOME not set"
+    exit 1
+fi
+
+if [ "x" == "${HADOOP_CONF_DIR}x" ]
+then
+    source //etc/profile.d/hadoop-env.sh
```
Comment (sookim-altiscale) [inline etc/init.d/hadoop_datanode:61]: Do you need double slash for root directory? (//etc/profile.d/hadoop-env.sh)

File: etc/init.d/hadoop_datanode [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+fi
+
+
+RETVAL=0
+# USER is required by hadoop-daemon.sh
+export USER=$HDFS_USER
+SERVICE="datanode"
+SERVICENAME="DataNode"
+DESC="Hadoop ${SERVICE} daemon"
+HADOOP_PID_DIR=/var/hadoop/hdfs/run/
```
Comment (sookim-altiscale) [inline etc/init.d/hadoop_datanode:88]: I think it's better to create the HADOOP_PID_DIR if that does not exist on the system.
Chef-solo has a responsibility to create the directory. This responsibility should be moved in init.d.

File: etc/init.d/hadoop_datanode [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+
+# HADOOP_DATANODE_OPTS is used by ${HADOOP_HOME}/bin/hdfs
+# which is called from ${HADOOP_HOME}/sbin/hadoop-daemon.sh
+# which is called by this initialization script.
+# Note that ${HADOOP_HOME}/sbin/hadoop-daemon.sh sources
+# ${HADOOP_CONF_DIR}/hadoop-env.sh
+# bin/hdfs does not yet implement a HEAPSIZE option.
+export HADOOP_SECURITY_LOGGER="${HADOOP_SECURITY_LOGGER:-INFO,DRFAS}"
+export HADOOP_DATANODE_OPTS="$DATANODE_OPTS\
+ -Xloggc:${HDFS_LOG_DIR}/hadoop-${USER}-${SERVICE}-gc.log-`date +'%Y%m%d%H%M'`\
```
Comment (sookim-altiscale) [inline etc/init.d/hadoop_datanode:108]: If HDFS_LOG_DIR does not exist, what happen? When you check the line 70, it also assumes the HDFS_LOG_DIR has a value.

**Comment by sookim-altiscale** [comment]
I agree your point we need to keep static init.d script.
I tried to review the code. However, I still don't know the scope of the this task.

A. Previous installation steps
a.1 install RPM
a.2 Using the chef-solo, we set the configuration.
First, default configuration(e.g. create directories, copy default files and create init.d for each daemon) 
Second, custom configuration(e.g. core-site.xml)

B. Feature installation steps to use chef server
b.1 install RPM (It has init.d)
b.2 Using the chef-server, we set the custom configuration.

My question is if we install the hadoop using this RPM, can I run the hadoop daemon using this default RPM configuration?
I think the goal of this task is to provide the default configuration which is used on init.d for Hadoop service RPM. Some of the default configuration is set by chef-solo right now.
Since I don't know the goal of this task, I am confused the depth of the code review.

**Review by sookim-altiscale — COMMENTED** [review]
I agree your point we need to keep static init.d script.
I tried to review the code. However, I still don't know the scope of the this task.

A. Previous installation steps
a.1 install RPM
a.2 Using the chef-solo, we set the configuration.
First, default configuration(e.g. create directories, copy default files and create init.d for each daemon)
Second, custom configuration(e.g. core-site.xml)

B. Feature installation steps to use chef server
b.1 install RPM (It has init.d)
b.2 Using the chef-server, we set the custom configuration.

My question is if we install the hadoop using this RPM, can I run the hadoop daemon using this default RPM configuration?
I think the goal of this task is to provide the default configuration which is used on init.d for Hadoop service RPM. Some of the default configuration is set by chef-solo right now.
Since I don't know the goal of this task, I am confused the depth of the code review.

File: etc/init.d/hadoop_datanode (hunk 3 above)
Comment (zoran-altiscale) [reply etc/init.d/hadoop_datanode:108]: guys, this change is 2181 lines - I don't expect it to be perfect and I don't want to change semantics of what these scripts are supposed to be doing.

So, please, keep the comments about how to refactor init.d scripts to be better for next change. I did not write these scripts and I cannot change the semantic in one PR.

The goal of this change is to refactor existing chef-solo init.d scripts into static code (in init.d/) and settings (in /etc/sysconfig/ dir). We plan to only use these new scripts (that all have new names) in chef server code, and when we launch particular service in chef server, then the old chef solo code becomes obsolete.

**Comment by raviprak-altiscale** [comment]
Ok! Given the purpose of this PR is not to fix the init scripts themselves, but in fact how they are installed on nodes (chef-solo earlier but RPMs now), its unfortunate that Github is showing the diff this way.

**A more accurate representation of the change** would be diffing e.g. https://github.com/VertiCloud/chef/blob/master/cookbooks/local/hadoop/templates/default/datanode-initd.erb and https://github.com/Altiscale/hadoopbuild/blob/zoran-HD-1878-initd/etc/init.d/hadoop_datanode (and similarly for the other files)

In that respect, this is a good change. The only thing I did notice is that we removed this line
```unless node['virtualization']['system'] == 'lxc' && node['virtualization']['role'] == 'guest' %>```
I suspect that's because we don't have our docker virtualized cluster (I can't remember its name) anymore.

I checked datanode, namenode and secondarynamenode. I trust the other files are also similar.

**Comment by raviprak-altiscale** [comment]
LGTM. +1

**Pull request merged**